### ISSUE TITLE
Replacing system malloc at startup with jemalloc on iOS

### ISSUE THREAD

**Issue by michaeleisel** [jemalloc/jemalloc#1322]
It would be great to replace `malloc` with with `jemalloc`. From my performance tests, jemalloc on iOS looks excellent. In order to do this, https://github.com/jemalloc/jemalloc/blob/dev/src/zone.c looks promising, but also unsafe, considering how many changes there are for specific OS X versions. We could use dyld interposing or something in a C constructor function, but this isn't early enough to catch all the places where the linker (and perhaps other libraries?) have invoked `malloc`. So the linker tries to free this pointer afterwards by calling jemalloc's `free`, and a crash occurs. I also wonder if there could be other concurrent operations during the replacement, and if that would cause any race conditions. We could also figure out whatever the address sanitizer does, since it replaces malloc at startup, but I couldn't figure it out myself.

Any thoughts?

**Comment by michaeleisel** [comment]
FYI we can't set any environment variables when the user runs the app, so we can't use things like LD_PRELOAD

**Comment by davidtgoldblatt** [comment]
Yeah, this is tricky; fundamentally I think we need some degree of libc support (or at least the lack of hostility) in order to make this work; it's not a given that it's possible.

I'm not sure what address sanitizer does; that might be a promising avenue, though.

**Comment by victorstewart** [comment]
any further developments along this trajectory?

**Comment by michaeleisel** [comment]
the approach used by zone.c has worked pretty well for me on ios

**Comment by victorstewart** [comment]
@michaeleisel 

kk i'll try that out. I'd wanted to go with interposing after reading this quote in the below, but definitely feel better about going with an existing implementation.

> "From speaking with an Apple engineer, they recommended copying address sanitizer's interposing method, and said that anything else is probably less safe."

[https://www.reddit.com/r/iOSProgramming/comments/9t3a9n/replacing_the_ios_system_allocator_for_fun_and/](https://www.reddit.com/r/iOSProgramming/comments/9t3a9n/replacing_the_ios_system_allocator_for_fun_and/)

**Comment by michaeleisel** [comment]
yeah, i should probably update that blog post haha

**Comment by aditya7fb** [comment]
I've had some success with using jemalloc. Happy to answer any questions.

**Comment by michaeleisel** [comment]
What have you seen in terms of its performance?

**Comment by aditya7fb** [comment]
Crashes due to OOMs reduced, and warm start perf improved.

**Comment by michaeleisel** [comment]
Very cool! How did you integrate it? Were there any stability issues? Are you afraid of any potential ones in the future as Apple releases new versions of libmalloc?

**Comment by aditya7fb** [comment]
> How did you integrate it? Were there any stability issues?
I'll share details soon.

> Are you afraid of any potential ones in the future as Apple releases new versions of libmalloc?
Unlikely as Apple themselves use multiple allocators AFAICT.

**Comment by aditya7fb** [comment]
Scripts to build: https://github.com/jemalloc/jemalloc/pull/2000

**Comment by aditya7fb** [comment]
and #2001 for conditionally enabling jemalloc for testing.

**Comment by michaeleisel** [comment]
Are you able to share what % decrease you saw in warm start time, or in OOMs?

**Comment by michaeleisel** [comment]
Also, I wonder if it'd make sense to call `zone_register` not as an initializer but instead explicitly a little later on after the crash reporter is setup. It wouldn't affect stability, right?

**Comment by gaxlin** [comment]
> I've had success with using jemalloc on ios. Happy to answer any questions.

How? could you write some guides?

**Comment by davidtgoldblatt** [comment]
The stuff Aditya did is in #2000 and #2001. If I remember correctly, he also had to pre-configure using the jemalloc build system with some known defaults, and then import the resulting files as a batch into the iOS build system, linked into a single blob.

(He's no longer working on this, btw).
